### ISSUE TITLE
apd: Identical APD numbers with and without trailing zeroes are considered different inside dataflow

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#6912]
### What version of Materialize are you using?

```
4e97d3867
```

### What was the issue?

I am opening this separately so as to not clutter the discussion about #6877 any further. 

On both "main" and on a branch with all the fixes for #6877, (branch ```apd-distinct```, rust-dec hash ```#42d24e33109d0187c06caf817c4cd7a577a07c26```, the numbers ```'0.1'::apd``` and ```0.10:apd``` are considered different.

### Is the issue reproducible? If so, please provide reproduction instructions.

```
DROP TABLE IF EXISTS deletions CASCADE ;
CREATE TABLE insertions (f1 APD) ;
CREATE TABLE deletions (f1 APD) ;
INSERT INTO insertions VALUES ( '0.1'::APD );
INSERT INTO deletions VALUES ( '0.10'::APD );
SELECT f1 FROM insertions EXCEPT ALL SELECT f1 FROM deletions;

 f1  
-----
 0.1 <- wrong result, should have returned no rows
(1 row)

```
Postgres aggrees.

**Comment by sploiselle** [comment]
This issue is caused by an impedence mismatch between our new decimal type's notion of equality and all other types'.

The underlying library that powers the decimal implmentation has a default notion of equality similar to all of our other types', i.e. byte-wise equality where `0.90` and `0.9` are not equal to one another.

However, to match Postgres' semantics, we needed to add feature that enabled us to express equality irrespective of values' trailing zeroes, i.e. `0.90` should equal `0.9`. We achieve this by using the `OrderedDecimal` struct in the library, which performs a `reduce` operation (i.e. trims all trailing zeroes) during every comparison.

_However_, no other types in Materialize use non-byte-wise equality, which means that our current invariants for row equality used in e.g. `DISTINCT`, `EXCEPT`, et. al., do not naturally/immediately apply to the new decimal type.

To solve this, we have a few different options that I see as immediately available/viable:

### No longer do byte-wise comparison for `Row` equality

Because not all types rely on byte-wise equality, it's possible that we no longer want to use it as the method to determine rows' equality, and instead iterate over the rows' data for comparisons.

The upside to this is its simplicity, while the downside is a considerable degradation in performance. I don't see this as a reasonable solution.

### Require fixed scale

`rust-dec` natively supports operations on values of both heterogeneous precision and scale; so far we've been able to preserve the arbitrariness of scale on operations, but that "feature" is essentially the impedence mismatch between the library and our current implementation of rows' equality.

One potential fix then is to require an explicit scale on all decimal values (we can offer a default of e.g. 0 or 4).

The downside here is that values lose a lot of their flexibility, i.e. you will no longer get 39 digits of precision anywhere it happens to occur, but will instead be working with some fixed number of values in the "integer" and "fractional" places. The up side to this is that it closely mirrors our current approach.

### Reduce all values

The real trouble with equality in the new `decimal` type is that we want to consider two values equal if their values after a `reduce` are equal, i.e. if they are the same modulo any trailing zeroes.

This means that we could instead call `reduce` on values before putting them into rows.

The upside to this is that it retains flexibility and can compress the size of the values we store. The down side to this is that, in contexts where a scale is not specified, trailing zeroes on values will disappear. Trailing zeroes are mostly cosmetic, but it's certainly odd to have them  disappear even in places you think they belong.

An additional up side to this change is that it's accomplishable with a single line of code (i.e. a call to `reduce` when deserializing a decimal value into a row).

After tinkering around with this for the morning, I'm inclined toward this approach.

**Comment by sploiselle** [comment]
@philip-stoev: Nikhil suggested simply doing the `reduce` operation before packing rows. I've pushed a commit that does that to my fork on the `apd-distinct` branch. Really appreciate your fastidiousness in continuing to test these changes despite the oddity of this set up. Will get these things in a state easier to work with in the next day or two.

**Comment by philip-stoev** [comment]
@sploiselle: Your commit does fix the above example with a single trailing zero, along with the case of two trailing zeroes. Unfortunately however three or more trailing zeroes are still considered distinct:

```
 DROP TABLE IF EXISTS insertions CASCADE ;
 DROP TABLE IF EXISTS deletions CASCADE ;
 CREATE TABLE insertions (f1 APD) ;
  CREATE TABLE deletions (f1 APD) ;
 INSERT INTO insertions VALUES ('0.1'::APD ) ;
INSERT INTO deletions VALUES ( '0.1000'::APD ) ;
SELECT f1 FROM (SELECT * FROM insertions EXCEPT ALL SELECT * FROM deletions ) AS a1;

 f1  
-----
 0.1
(1 row)
```

**Comment by sploiselle** [comment]
Interesting! I had only tested this with `distinct` (assuming it used the same code in some way):

```
drop table if exists a cascade;
create table a (f1 apd);
insert into a values ('.1'),('.10'),('.100'),('.1000'),('.10000');
select distinct f1 from a;
```

```
 f1
-----
 0.1
(1 row)
```

Will start digging into how `EXCEPT` differs.

**Comment by sploiselle** [comment]
Ah, ok––calls to `reduce` in the C library do not remove truncated digits from the interior buffer; annoyingly, the type we're using doesn't provide a method to do that either so need to hand-roll something. Will report back here once I get that done.

**Comment by sploiselle** [comment]
So the root cause of this behavior is my implicit misunderstanding of a function called [`decShiftToLeast`](https://github.com/MaterializeInc/rust-dec/blob/master/decnumber-sys/decnumber/decNumber.c#L6726-L6786) in the underlying C library. This function left-shifts values in the digit buffer after calls to e.g. `reudce`, and adjusts the `digits` field, but does not "zero out" any digits that it shifted that now lay beyond the `digits` barrier. This means a call to `reduce` makes everything look like you'd expect, but leaves some values within the `lsu`––this is cause of the issue we're running into. (I'm not sure how this gets elided in the `DISTINCT` case).

To solve this, we can either:
- Allocate a new value set to zero, and then add the existing value to it; this has the side effect of "zeroing out" the `lsu`. The up side is that this uses existing functionality, but the down side is that it will perform an additional allocation for each value each time it's packed into a row.
- Add a new (probably `unsafe`?) function that cleans the `lsu` of any extraneous values on demand. The up side is that this avoids an unnecessary allocation; the down side is that we're manually manipulating memory which isn't great.

@benesch You have an opinion/see other reasonable options?

**Comment by benesch** [comment]
Yeah, per #6883, we'll want to only write down the part of the LSU that's valid according to the digits field to keep things small, and it will also solve this issue!

Oh, also, one minor point of pedantry: `Decimal<N>` is not heap allocated. That array gets allocated on the stack!

**Comment by sploiselle** [comment]
@benesch ty for the correction; I had been cargo culting a notion I thought I heard that we were allocating values without giving it much though but realize now that is untrue.

**Comment by sploiselle** [comment]
Closed by #6945
